tls/s2n_psk.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "crypto/s2n_tls13_keys.h"
+#include "tls/s2n_handshake.h"
+#include "tls/s2n_tls13_handshake.h"
+#include "tls/s2n_psk.h"
+#include "utils/s2n_safety.h"
+#include "utils/s2n_mem.h"
+
+int s2n_psk_init(struct s2n_psk *psk, s2n_psk_type type)
+{
+    notnull_check(psk);
+
+    memset_check(psk, 0, sizeof(struct s2n_psk));
+    psk->hash_alg = S2N_HASH_SHA256;
+    psk->type = type;
+
+    return S2N_SUCCESS;
+}
+
+int s2n_psk_new_identity(struct s2n_psk *psk, const uint8_t *identity, size_t identity_size)
+{
+    notnull_check(psk);
+
+    GUARD(s2n_realloc(&psk->identity, identity_size));
```

Comment:
Q: Why not use `s2n_alloc` here? `s2n_alloc` would set the data, size and allocated to 0 (as required  by _new) and calls `s2n_realloc`.

Reply:
I'm using s2n_realloc to handle the case where this method is called repeatedly.

s2n_alloc is just a wrapper that zero-initializes the blob before calling s2n_realloc. If psk->indentity has already been alloc'd, zero-initializing before calling s2n_realloc would lead to a memory leak.

Reply:
But why would psk->Identity be alloc'd prior to calling this function? Isn't this the function we would like to alloc the memory for the psk->identity?

Reply:
There is no mechanism to prevent anyone from calling this method twice. If you used alloc, you would need to specifically check that the function was not called twice and either error or clean up psk->identity if it is. It's simpler just to let it happen and let realloc deal with it. There's no reason not to use realloc.